system/lib/pthread/library_pthread.c

```diff
@@ -192,49 +193,103 @@ void emscripten_async_waitable_close(em_queued_call *call)
+			case EM_FUNC_SIG_I: q->returnValue.i = _emscripten_pthread_do_call_js_i(q->js); break;
+			case EM_FUNC_SIG_D: q->returnValue.d = _emscripten_pthread_do_call_js_d(q->js); break;
+			case EM_FUNC_SIG_VI: _emscripten_pthread_do_call_js_vi(q->js, q->args[0].i); break;
+			case EM_FUNC_SIG_II: q->returnValue.i = _emscripten_pthread_do_call_js_ii(q->js, q->args[0].i); break;
+			case EM_FUNC_SIG_FI: q->returnValue.f = _emscripten_pthread_do_call_js_fi(q->js, q->args[0].i); break;
+			case EM_FUNC_SIG_DI: q->returnValue.d = _emscripten_pthread_do_call_js_di(q->js, q->args[0].i); break;
+			case EM_FUNC_SIG_VII: _emscripten_pthread_do_call_js_vii(q->js, q->args[0].i, q->args[1].i); break;
+			case EM_FUNC_SIG_VIF: _emscripten_pthread_do_call_js_vif(q->js, q->args[0].i, q->args[1].f); break;
+			case EM_FUNC_SIG_III: q->returnValue.i = _emscripten_pthread_do_call_js_iii(q->js, q->args[0].i, q->args[1].i); break;
+			case EM_FUNC_SIG_VIII: _emscripten_pthread_do_call_js_viii(q->js, q->args[0].i, q->args[1].i, q->args[2].i); break;
+			case EM_FUNC_SIG_VIIF: _emscripten_pthread_do_call_js_viif(q->js, q->args[0].i, q->args[1].i, q->args[2].f); break;
+			case EM_FUNC_SIG_IIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i); break;
+			case EM_FUNC_SIG_VIIII: _emscripten_pthread_do_call_js_viiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i); break;
+			case EM_FUNC_SIG_VIFFF: _emscripten_pthread_do_call_js_vifff(q->js, q->args[0].i, q->args[1].f, q->args[2].f, q->args[3].f); break;
+			case EM_FUNC_SIG_IIIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i); break;
+			case EM_FUNC_SIG_IIFFF: q->returnValue.i = _emscripten_pthread_do_call_js_iifff(q->js, q->args[0].i, q->args[1].f, q->args[2].f, q->args[3].f); break;
+			case EM_FUNC_SIG_IIIIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i); break;
+			case EM_FUNC_SIG_IIIIIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiiiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i); break;
+			case EM_FUNC_SIG_VIIIIII: _emscripten_pthread_do_call_js_viiiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i); break;
+			case EM_FUNC_SIG_IIIIIIIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiiiiiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i); break;
+			case EM_FUNC_SIG_IIIIIIIIII: q->returnValue.i = _emscripten_pthread_do_call_js_iiiiiiiiii(q->js, q->args[0].i, q->args[1].i, q->args[2].i, q->args[3].i, q->args[4].i, q->args[5].i, q->args[6].i, q->args[7].i, q->args[8].i); break;
+			default: assert(0 && "Invalid Emscripten pthread _do_call (js) opcode!");
+		}
```

Comment:
The generated code size of all these signature switch tables is a big concern. Theoretically it feels that we should be able to get by with having exactly one signature switch table in C/C++ side, and at most one signature switch table in JS side - both of which would eventually be dynamically generated by final link step (or if dynamic linking with unknown code possibly loaded in, a default supported set of signatures would be generated). I wonder if we could get rid of this `else` path and merge this part to above by assigning an appropriate `q->functionPtr` at the proxy call site, and perhaps add the JS function index as the first parameter. Then in here we could always go through the `if()` part above without having to duplicate all signatures.

Overall it feels that the O(#signatures) combinatorial complexity should occur at most once in JS and in C/C++ side. 

Also something that might be possible to do is that we could reduce the combinatorial explosion of JS proxied functions by always treating their return value and all parameters to be doubles - that would collapse the type combinatorics down by quite a lot. That way we could treat the parameters and return values as variadic on the heap, and instead of having an `EM_JS()` block for each signature, we could reuse the same kind of pattern that currently exists in `library_pthread.js worker.onmessage block:`

```js
            if (d.proxiedCall) {
              var returnValue;
              var funcTable = (d.func >= 0) ? proxiedFunctionTable : ASM_CONSTS;
              var funcIdx = (d.func >= 0) ? d.func : (-1 - d.func);
              switch(d.proxiedCall & 31) {
                case 1: returnValue = funcTable[funcIdx](); break;
                case 2: returnValue = funcTable[funcIdx](d.p0); break;
                case 3: returnValue = funcTable[funcIdx](d.p0, d.p1); break;
                case 4: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2); break;
                case 5: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3); break;
                case 6: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3, d.p4); break;
                case 7: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3, d.p4, d.p5); break;
                case 8: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3, d.p4, d.p5, d.p6); break;
                case 9: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3, d.p4, d.p5, d.p6, d.p7); break;
                case 10: returnValue = funcTable[funcIdx](d.p0, d.p1, d.p2, d.p3, d.p4, d.p5, d.p6, d.p7, d.p8); break;
                default:
                  if (d.proxiedCall) {
                    err("worker sent an unknown proxied call idx " + d.proxiedCall);
                    console.error(e.data);
                  }
                  break;
              }
```

which is very lean - no O(#signatures) explosion but only O(#argumentCount). Perhaps we could have a JS library function with signature `double emscripten_run_proxied_call(int functionIndex, double* arguments, int numArguments);` that is code generated to have a switch-case for each argument count known to the compiler that will be proxied. Then we could manage that as a `EM_FUNC_SIG_DII` function and place the address of `emscripten_run_proxied_call()` to `q->functionPtr`, so then we would not need a `q->js` field?

Reply:
Yeah, good points here. I think we can actually use the dynamic nature to JS here to avoid any switch at all, in fact, which would be simpler and much better for code size, and probably not add any noticeable runtime overhead anyhow.